POLIO-1597 Chronogram Frontend

kemar opened the pull request:
Chronogram Frontend.

Related JIRA tickets : [POLIO-1597](https://bluesquare.atlassian.net/browse/POLIO-1597)

## Doc

https://whimsical.com/chronograms-X1VchmEGLneHBFiQbZcTfy

## Changes

- new UIs for Chronograms
    1. list Chronograms
    1. CRUD for Chronograms task templates (these tasks are automatically appended to new Chronograms)
    1. CRUD for Chronograms tasks
- backend bug fixes
- add backend translations for Polio (it didn't exist until now)

## How to test

Go to `/dashboard/polio/chronogram/` and test the UI.

## Video

https://github.com/BLSQ/iaso/assets/281139/2a91f2aa-76a3-4a87-8f04-18e505b1bf7f




[POLIO-1597]: https://bluesquare.atlassian.net/browse/POLIO-1597?atlOrigin=eyJpIjoiNWRkNTljNzYxNjVmNDY3MDlhMDU5Y2ZhYzA5YTRkZjUiLCJwIjoiZ2l0aHViLWNvbS1KU1cifQ

kemar committed:
Add default route and perms for Chronogram

kemar committed:
Add chronogram list

kemar committed:
Add chronogram template task list

kemar committed:
Move Chronogram code into its own directory

kemar committed:
Use `goBack`

kemar committed:
Allow to filter ChronogramTasks by `chronogram_id`

kemar committed:
Add display values to Chronogram serializers

kemar committed:
Add chronogram task list

kemar committed:
Allow to delete a default chronogram task

kemar committed:
Filter on valid objects

kemar committed:
Allow to delete a chronogram task

kemar committed:
Fix messages

kemar committed:
Set `account` automatically in `ChronogramTemplateTaskSerializer`

kemar committed:
Allow to add and edit a default chronogram task

kemar committed:
Allow to add and edit chronogram tasks

kemar committed:
Use only `.valid()` objects

kemar committed:
Allow to add a chronogram

kemar committed:
Add JS translations

kemar committed:
Use `NumberInput`

kemar committed:
Use `get_serializer_class` to display a proper POST form in DRF UI

kemar committed:
Use `gettext_lazy` for translations

kemar committed:
Add Polio translations

beygorghor commented on plugins/polio/js/src/domains/Chronogram/Chronogram/Filters/ChronogramFilters.tsx:
```diff
@@ -0,0 +1,110 @@
+
+    const { countriesData, isFetchingCountriesData: isFetchingCountries } =
+        useGetCountries();
+    const countriesOptions = (countriesData && countriesData.orgUnits) || [];
```
For this you can use the select option from the react-query hook

kemar replied:
Can you please show me how to do that or point me to a place in the code where it's done?

beygorghor commented on plugins/polio/js/src/domains/Chronogram/Chronogram/Filters/ChronogramFilters.tsx:
```diff
@@ -0,0 +1,110 @@
+        useGetCountries();
+    const countriesOptions = (countriesData && countriesData.orgUnits) || [];
+
+    const onTimeOptions = [
+        {
+            label: formatMessage(MESSAGES.yes),
+            value: 'true',
+        },
+        {
+            label: formatMessage(MESSAGES.no),
+            value: 'false',
+        },
+    ];
```
may create a separate hook for this or use a useMemo to avoid recomputing on each render

beygorghor commented on plugins/polio/js/src/domains/Chronogram/Chronogram/Modals/CreateChronogramModal.tsx:
```diff
@@ -0,0 +1,133 @@
+    const allowConfirm = !isSubmitting && isValid && dirty;
+
+    // Filter "Campaign" values on "Country" change.
+    const [currentCampaignOptions, setCampaignOptions] = useState<
+        DropdownOptions[]
+    >([]);
+    useEffect(() => {
+        const campaigns = dropdownsData?.campaigns || [];
+        const filtered = campaigns.filter(
+            i => String(i.country_id) === String(formik.values.country),
+        );
+        formik.setFieldValue('campaign', '');
+        setCampaignOptions(filtered);
+        // eslint-disable-next-line react-hooks/exhaustive-deps
+    }, [dropdownsData?.campaigns, formik.values.country]);
```
you can replace the useEffect with a useMemo to get currentCampaignOptions, now i d'ont understand the setFieldValue here

kemar replied:
Can you please explain me the difference between `useEffect` and `useMemo` and tell me why `useMemo` would be better here?

kemar replied:
> now i d'ont understand the setFieldValue here

When the country is changed, we need to reset the campaign field. This is why I used `setFieldValue`.

beygorghor commented on plugins/polio/js/src/domains/Chronogram/Chronogram/Modals/CreateChronogramModal.tsx:
```diff
@@ -0,0 +1,133 @@
+    }, [dropdownsData?.campaigns, formik.values.country]);
+
+    // Filter "Rounds" values on "Campaign" change.
+    const [currentRoundsOptions, setRoundsOptions] = useState<
+        DropdownOptions[]
+    >([]);
+    useEffect(() => {
+        const rounds = dropdownsData?.rounds || [];
+        const filtered = rounds.filter(
+            i => String(i.campaign_id) === String(formik.values.campaign),
+        );
+        formik.setFieldValue('round', undefined);
+        setRoundsOptions(filtered);
+        // eslint-disable-next-line react-hooks/exhaustive-deps
+    }, [dropdownsData?.rounds, formik.values.campaign]);
```
same here with currentRoundsOptions

beygorghor commented on plugins/polio/js/src/domains/Chronogram/Chronogram/Modals/CreateChronogramModal.tsx:
```diff
@@ -0,0 +1,133 @@
+        // eslint-disable-next-line react-hooks/exhaustive-deps
+    }, [dropdownsData?.rounds, formik.values.campaign]);
+
+    const titleMessage = formatMessage(MESSAGES.createChronogramTitle);
```
this is useless

beygorghor commented on plugins/polio/js/src/domains/Chronogram/Chronogram/Table/ChronogramTable.tsx:
```diff
@@ -0,0 +1,41 @@
+    params,
+    chronogramTaskMetaData,
+}) => {
+    const apiParams: ChronogramParams = {
+        ...params,
+        limit: params.pageSize || '20',
+        order: params.order || 'id',
+        page: params.page || '1',
+    };
```
useMemo

kemar replied:
why?

beygorghor commented on plugins/polio/js/src/domains/Chronogram/Chronogram/api/useGetChronogram.ts:
```diff
@@ -0,0 +1,27 @@
+    params: ChronogramParams,
+): UseBaseQueryResult<ChronogramApiResponse, unknown> => {
+    // Removed params with an undefined value.
+    const cleanedParams: Partial<ChronogramParams> = JSON.parse(
+        JSON.stringify(params),
+    );
```
Not a big fan of Json parse, please use URLSearchParams that will remove undefined values

kemar replied:
AFAIK `URLSearchParams` doesn't remove undefined values:

```javascript
> const params = {order: "id", search: undefined}

> new URLSearchParams(params);
URLSearchParams { 'order' => 'id', 'search' => 'undefined' }

> JSON.parse(JSON.stringify(params));
{ order: 'id' }
```

beygorghor commented on plugins/polio/js/src/domains/Chronogram/ChronogramDetails/api/useGetChronogramTasks.ts:
```diff
@@ -0,0 +1,27 @@
+    params: ChronogramTasksParams,
+): UseBaseQueryResult<ChronogramTaskApiResponse, unknown> => {
+    // Removed params with an undefined value.
+    const cleanedParams: Partial<ChronogramTasksParams> = JSON.parse(
+        JSON.stringify(params),
+    );
```
same here

beygorghor commented on plugins/polio/js/src/domains/Chronogram/ChronogramDetails/index.tsx:
```diff
@@ -0,0 +1,80 @@
+
+    return (
+        <>
+            {isFetching && isFetchingMetaData && <LoadingSpinner />}
+            {!isFetching && !isFetchingMetaData && (
```
i would place the loader under the topbar to be consistent with the rest of the application

beygorghor commented on plugins/polio/js/src/domains/Chronogram/ChronogramTemplateTask/api/useGetChronogramTemplateTask.ts:
```diff
@@ -0,0 +1,30 @@
+    params: ChronogramTemplateTaskParams,
+): UseBaseQueryResult<ChronogramTemplateTaskApiResponse, unknown> => {
+    // Removed params with an undefined value.
+    const cleanedParams: Partial<ChronogramTemplateTaskParams> = JSON.parse(
+        JSON.stringify(params),
+    );
```
URLSearchParams

beygorghor commented on plugins/polio/js/src/domains/Chronogram/ChronogramTemplateTask/index.tsx:
```diff
@@ -0,0 +1,74 @@
+        ...params,
+        pageSize: params.pageSize || '20',
+        order: params.order || 'id',
+        page: params.page || '1',
```
maybe use a default pagination params, this is not the first time you make this

beygorghor commented on plugins/polio/js/src/domains/Chronogram/ChronogramTemplateTask/index.tsx:
```diff
@@ -0,0 +1,74 @@
+
+    return (
+        <>
+            {isFetchingMetaData && <LoadingSpinner />}
+            {!isFetchingMetaData && (
```
same here

beygorghor commented on plugins/polio/js/src/domains/Chronogram/api/useOptionChronogramTask.ts:
```diff
@@ -0,0 +1,32 @@
+            cacheTime: 1000 * 60 * 5,
+            select: data => {
+                const metadata = data.actions.POST;
+                const mapChoices = choices =>
+                    choices.map(choice => ({
+                        label: choice.display_name,
+                        value: choice.value,
+                    }));
```
You can create a useCallback function with this no ?

kemar replied:
I don't know how to do that, can you show me please?

beygorghor reviewed (changes requested):
Wow nice job ! Very clean 

- I had a 500 while creating a new chronogram and trying to get the list.
Not sure this is not due to the quality of my data:
<img width="686" alt="Screenshot 2024-07-11 at 11 05 07" src="https://github.com/BLSQ/iaso/assets/12494624/ab459950-7ec9-47c4-8bac-de9693ccb1ca">
- This Dialog is too large:
<img width="1115" alt="Screenshot 2024-07-11 at 11 04 10" src="https://github.com/BLSQ/iaso/assets/12494624/890e1ea4-ac68-49ba-9017-1ef79b72aba5">
- Not sure we need the select visible columns on this page
<img width="1893" alt="Screenshot 2024-07-11 at 11 03 54" src="https://github.com/BLSQ/iaso/assets/12494624/c5bd2084-2821-4384-87e2-9b66c1c81780">
- Maybe add infos with InputWithInfos to explain what is start for 
<img width="738" alt="Screenshot 2024-07-11 at 11 03 31" src="https://github.com/BLSQ/iaso/assets/12494624/4bda4d0e-3d34-4f61-b0ce-54794c2e564e">
- some remarks in the code too

kemar committed:
Prevent a Chronogram creation if the linked round doesn't have a `started_at` value

kemar committed:
Remove useless variable

kemar committed:
Place the loading consistently with the rest of the app

kemar committed:
Use a default pagination params

kemar commented:
> * I had a 500 while creating a new chronogram and trying to get the list.
>   Not sure this is not due to the quality of my data:

I've fixed that. It happened for rounds without start date. I added a validation rule to prevent chronogram creation for rounds without start date.

kemar committed:
Use `xs` for the width of the modal used to create chronograms

kemar commented:
> - This Dialog is too large:

Fixed.

kemar committed:
Disable column selector on chronogram default tasks list

kemar commented:
> - Not sure we need the select visible columns on this page

Fixed.

kemar committed:
Use annotations instead of properties on `ChronogramTask`

kemar committed:
Allow to sort by `delay` and `deadline`

kemar committed:
Use annotations instead of properties on `Chronogram`

kemar committed:
Allow to sort by `num_task_delayed` and `is_on_time`

kemar committed:
Merge pull request #1479 from BLSQ/POLIO-1612-chronogram-perf

POLIO-1612 Chronogram - Backend optimizations
